*: add generic dialers to create RPC clients

cthumuluru-crdb opened the pull request:
This PR has changes to move `ExperimentalDRPCEnabled` flag to `rpcbase` package and add generic dial methods to create RPC clients. This refactor reduces the number of places where the DRPC experimental flag must be enabled, and simplifies adding version guards by limiting them to only a few locations.

Epic: CRDB-48935
Informs: #153032
Release note: None

cthumuluru-crdb committed:
rpc,server,testutils: move DRPC cluster setting to rpcbase

Epic: CRDB-48935
Informs: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/152948)
<!-- Reviewable:end -->

cthumuluru-crdb committed:
rpc: add generic dial methods to dial RPC clients

Introduce generic dial methods for creating RPC clients across services. This
consolidates code paths, reduces the number of places where the DRPC
experimental flag must be enabled, and simplifies adding version guards by
limiting them to only a few locations.

Epic: CRDB-48935
Informs: None
Release note: None

shubhamdhama reviewed (approved):
Looks good (with a suggestion/question).

shubhamdhama commented on pkg/rpc/nodedialer/nodedialer.go:
```diff
@@ -430,3 +430,32 @@ func (c *tracingInternalClient) Batch(
 	return c.RestrictedInternalClient.Batch(ctx, ba)
 }
+
+// DialRPCClient establishes a connection to a node identified by its ID and
+// returns a client for the requested service type. When DRPC is enabled, it
+// creates a DRPC client; otherwise, it falls back to a gRPC client.
+func DialRPCClient[C any](
+	nd *Dialer,
+	ctx context.Context,
+	nodeID roachpb.NodeID,
+	class rpcbase.ConnectionClass,
+	grpcClientFn func(*grpc.ClientConn) C,
+	drpcClientFn func(drpc.Conn) C,
+) (C, error) {
+	return rpcbase.DialRPCClient(nd, ctx, nodeID, class, grpcClientFn,
+		drpcClientFn, nd.rpcContext.Settings)
+}
+
+// DialRPCClientNoBreaker is like DialRPCClient, but will not check the
+// circuit breaker before trying to connect.
+func DialRPCClientNoBreaker[C any](
+	nd *Dialer,
+	ctx context.Context,
+	nodeID roachpb.NodeID,
+	class rpcbase.ConnectionClass,
+	grpcClientFn func(*grpc.ClientConn) C,
+	drpcClientFn func(drpc.Conn) C,
+) (C, error) {
+	return rpcbase.DialRPCClientNoBreaker(nd, ctx, nodeID, class, grpcClientFn,
+		drpcClientFn, nd.rpcContext.Settings)
+}
```
Oh this one also passes the settings object.
